flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecGroupWindowAggregateBase.scala

```diff
@@ -0,0 +1,345 @@
+        throw new TableException(
+          s"Group window $aggType must defined on a time attribute, " +
+            "but the time attribute can't be found.\n" +
+            "This should never happen. Please file an issue."
+        )
+      }
+      inputTimeFieldIndex
+    } else {
+      -1
+    }
+
+    val needRetraction = StreamExecRetractionRules.isAccRetract(getInput)
+    val aggInfoList = transformToStreamAggregateInfoList(
+      aggCalls,
+      inputRowType,
+      Array.fill(aggCalls.size)(needRetraction),
+      needInputCount = needRetraction,
+      isStateBackendDataViews = true)
+
+    val aggCodeGenerator = getAggsHandlerCodeGenerator(
+      aggInfoList,
+      config,
+      planner.getRelBuilder,
+      inputRowTypeInfo.getLogicalTypes,
+      needRetraction)
+
+    val aggResultTypes = aggInfoList.getActualValueTypes.map(fromDataTypeToLogicalType)
+    val windowPropertyTypes = namedProperties.map(_.property.resultType).toArray
+    val generator = new EqualiserCodeGenerator(aggResultTypes ++ windowPropertyTypes)
+    val equaliser = generator.generateRecordEqualiser("WindowValueEqualiser")
+
+    val aggValueTypes = aggInfoList.getActualValueTypes.map(fromDataTypeToLogicalType)
+    val accTypes = aggInfoList.getAccTypes.map(fromDataTypeToLogicalType)
+    val operator = createWindowOperator(
+      config,
+      aggCodeGenerator,
+      equaliser,
+      accTypes,
+      windowPropertyTypes,
+      aggValueTypes,
+      inputRowTypeInfo.getLogicalTypes,
+      timeIdx,
+      aggInfoList)
+
+    val transformation = new OneInputTransformation(
+      inputTransform,
+      getRelDetailedDescription,
+      operator,
+      outRowType,
+      inputTransform.getParallelism)
+
+    if (inputsContainSingleton()) {
+      transformation.setParallelism(1)
+      transformation.setMaxParallelism(1)
+    }
+
+    val selector = KeySelectorUtil.getBaseRowSelector(grouping, inputRowTypeInfo)
+
+    // set KeyType and Selector for state
+    transformation.setStateKeySelector(selector)
+    transformation.setStateKeyType(selector.getProducedType)
+    transformation
+  }
+
+  private def getAggsHandlerCodeGenerator(
+    aggInfoList: AggregateInfoList,
+    config: TableConfig,
+    relBuilder: RelBuilder,
+    fieldTypeInfos: Seq[LogicalType],
+    needRetraction: Boolean): AggsHandlerCodeGenerator = {
+
+    val needMerge = window match {
+      case SlidingGroupWindow(_, _, size, _) if hasTimeIntervalType(size) => true
+      case SessionGroupWindow(_, _, _) => true
+      case _ => false
+    }
+
+    val generator = new AggsHandlerCodeGenerator(
+      CodeGeneratorContext(config),
+      relBuilder,
+      fieldTypeInfos,
+      copyInputField = false)
+
+    generator.needAccumulate()
+    if (needMerge) {
+      generator.needMerge(mergedAccOffset = 0, mergedAccOnHeap = false)
+    }
+    if (needRetraction) {
+      generator.needRetract()
+    }
+
+    generator
+  }
+
+  protected def enrichWindowOperatorBuilder(
```

Comment:
I would like to finish the build logic in one place, rather than passing a half-work builder here and there. 
Can we use a if-else to generate the window operator in the base class depends on whether it is a table aggregate? like what we do in `DataStreamGroupWindowAggregateBase`?